The content of this file is derived from original patches published to libc-alpha at sourceware dot org
by Szabolcs Nagy. The original, unmodified, version is available in the patches/debian/2.27 folder of
this repository or from https://sourceware.org/legacy-ml/libc-alpha/2016-11/msg01092.html

The original content has been adapted by MathWorks to enable patching glibc 2.28 since there were 
source code changes in glibc since the original patches were constructed

NOTE: ADAPTED TO PATCH v2.28 via rpmbuild

From d66a4500aaf93d7a59b1ea668a60ff17c84d8d26 Mon Sep 17 00:00:00 2001
From: Szabolcs Nagy <szabolcs dot nagy at arm dot com>
Date: Wed, 30 Nov 2016 11:44:25 +0000
Subject: [PATCH 1/3] remove broken code path for easier code review

This patch is not necessary for the bug fix, just makes concurrency
code review easier (removes a data race and overflow from a broken
code path).

dlopen can oom crash anyway in _dl_resize_dtv and it's probably
better to crash than leave half setup modules around.

2016-11-30  Szabolcs Nagy  <szabolcs.nagy@arm.com>

	* elf/dl-tls.c (_dl_add_to_slotinfo): OOM crash.
---
 elf/dl-tls.c | 16 ++++------------
 1 file changed, 4 insertions(+), 12 deletions(-)

diff --git a/elf/dl-tls.c b/elf/dl-tls.c
index cccf74b3..14fe7c6f 100644
--- a/elf/dl-tls.c
+++ b/elf/dl-tls.c
@@ -998,18 +998,10 @@ _dl_add_to_slotinfo (struct link_map *l, bool do_add)
 		+ TLS_SLOTINFO_SURPLUS * sizeof (struct dtv_slotinfo));
       if (listp == NULL)
 	{
-	  /* We ran out of memory.  We will simply fail this
-	     call but don't undo anything we did so far.  The
-	     application will crash or be terminated anyway very
-	     soon.  */
-
-	  /* We have to do this since some entries in the dtv
-	     slotinfo array might already point to this
-	     generation.  */
-	  ++GL(dl_tls_generation);
-
-	  _dl_signal_error (ENOMEM, "dlopen", NULL, N_("\
-cannot create TLS data structures"));
+	  /* We ran out of memory in dlopen while updating tls structures.
+	     TODO: side-effects should be rolled back and the failure should
+	     be reported to the caller, but that's hard.  */
+	  oom ();
 	}
 
       listp->len = TLS_SLOTINFO_SURPLUS;
-- 
2.27.0

